tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java

```diff
@@ -374,6 +379,84 @@ public GraphOperationBuilder opBuilder(String type, String name) {
     return new GraphOperationBuilder(this, type, name);
   }
 
+  @Override
+  public void attachFunction(ConcreteFunction function) {
+    try (Reference ref = ref();
+        PointerScope scope = new PointerScope()) {
+      attachNativeFunction(ref.nativeHandle(), function.nativeHandle());
+      function.getDependencies().forEach(x -> attachNativeFunction(ref.nativeHandle(), x));
+    }
+  }
+
+  private void attachNativeFunction(TF_Graph graph, TF_Function fn) {
+    TF_Status status = TF_Status.newStatus();
+    TF_GraphCopyFunction(graph, fn, null, status);
+    status.throwExceptionIfNotOK();
+  }
+
+  synchronized List<NativeFunction> getNativeFunctions() {
```

Comment:
can you explain briefly why this (and the following) endpoints need to be synchronized? Looks like they are mainly a read operation and there is no synchronized-write operation paired with it

Reply:
Just to prevent someone from closing the graph while I'm getting functions.  Although it seems like that's handled by `ref`?  If that's the case I'll remove it.